cocos/renderer/gfx-vulkan/VKGPUObjects.h

```diff
@@ -470,6 +474,218 @@ class CCVKGPUCommandBufferPool : public Object {
     vector<uint> _counts;
 };
 
+class CCVKGPUStagingBufferPool : public Object {
+public:
+    struct Buffer {
+        VkBuffer vkBuffer = VK_NULL_HANDLE;
+        VkDeviceSize size = 16 * 1024 * 1024; // 16M per block by default
+        uint8_t *mappedData = nullptr;
+        VmaAllocation vmaAllocation = VK_NULL_HANDLE;
+
+        VkDeviceSize curOffset = 0u;
+    };
+
+    CCVKGPUStagingBufferPool(CCVKGPUDevice *device)
+    : _device(device) {
+    }
+
+    ~CCVKGPUStagingBufferPool() {
+        for (Buffer &buffer : _pool) {
+            vmaDestroyBuffer(_device->memoryAllocator, buffer.vkBuffer, buffer.vmaAllocation);
+        }
+        _pool.clear();
+    }
+
+    void alloc(CCVKGPUBuffer *gpuBuffer) {
+        size_t bufferCount = _pool.size();
+        Buffer *buffer = nullptr;
+        for (size_t idx = 0u; idx < bufferCount; idx++) {
+            Buffer *cur = &_pool[idx];
+            if (cur->size - cur->curOffset > gpuBuffer->size) {
+                buffer = cur;
+                break;
+            }
+        }
+        if (!buffer) {
+            _pool.resize(bufferCount + 1);
+            buffer = &_pool.back();
+            VkBufferCreateInfo bufferInfo{VK_STRUCTURE_TYPE_BUFFER_CREATE_INFO};
+            bufferInfo.size = buffer->size;
```

Comment:
should size be gpuBuffer->size?

Reply:
nope, the staging buffer is allocated in chunks, 16M per chunk for now